### PR TITLE
Updating rosetta dockerfile

### PR DESCRIPTION
**Pull request by wileyj**
- Refactoring to use runit vs custom script
- clone the latest releases vs static versions

## Description
This change refactors the rosetta Dockerfile to use [runit](http://smarden.org/runit/) as a scheduler vs a customized shell script. This change was motivated by discussions with another team trying to use this Dockerfile, and realizing the static version was very out of date. Further work was added to standardize the Dockerfile, so there weren't things like custom users/scripts (particularly in the case of postgresql). 

The *builder* images are also refactored to more closely resemble the default Dockerfiles, as well as cloning the latest release for each of `stacks-blockchain-api` and `stacks-blockchain` repos. 

Other changes are more use of ARG and ENV variables to create the scripts, i.e.
```
printf '#!/bin/sh
exec svlogd -tt %s/postgresql' ${STACKS_LOG_DIR} > ${STACKS_SVC_DIR}/postgresql/log/run
```

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] API reference/documentation update
- [x] Other

### CODE REVIEW

**Comment by wileyj** [comment]
ping @zone117x

File: stx-rosetta.Dockerfile
Comment (CharlieC3) [inline stx-rosetta.Dockerfile:69]: I think this is handled automatically now. @zone117x can you confirm?

File: stx-rosetta.Dockerfile
Comment (CharlieC3) [inline stx-rosetta.Dockerfile:6]: I think this ENV var can be removed. `${STACKS_API_REPO}` will just use the `ARG` like it's an ENV var. The ENV var directive should be used when the env variable needs to be available in the resulting image.

File: stx-rosetta.Dockerfile
Comment (CharlieC3) [inline stx-rosetta.Dockerfile:69]: Can we combine all of these `ENV` directives into one directive? It'll result in fewer layers in the image and decrease its size. Also if any of them aren't used in the final image's runtime, you can just use their `ARG` counterpart due to the same reason in my other comment.

File: stx-rosetta.Dockerfile
Comment (CharlieC3) [inline stx-rosetta.Dockerfile:103]: Have you considered committing the ruinit service files to this repo, and `COPY`ing them into the image? Might be easier to maintain and read that way.

File: stx-rosetta.Dockerfile
Comment (CharlieC3) [inline stx-rosetta.Dockerfile:70]: Combining this `RUN` command with the following three will further decrease the number of layers and size of the resulting image.

File: stx-rosetta.Dockerfile
Comment (wileyj) [reply stx-rosetta.Dockerfile:6]: unless the repo name changes, in which case this is a single change or a build-arg vs multiple changes in the file

File: stx-rosetta.Dockerfile
Comment (wileyj) [reply stx-rosetta.Dockerfile:69]: Yes, but it would also be incredibly hard to read. the goal here isn't a small image

File: stx-rosetta.Dockerfile
Comment (wileyj) [reply stx-rosetta.Dockerfile:103]: Possible, sure - but changes to the repo would require a full rebuild since the cache layers would no longer be used. 
Since there are many parts of this process, the goal was to create an image that can be built using cache layers as quickly as is feasible.

File: stx-rosetta.Dockerfile
Comment (wileyj) [reply stx-rosetta.Dockerfile:70]: Same as above - any change in the process/dockerfile would invalidate the cache layers and cause a longer rebuild. 
Can they all be combined? yes - iniitially that's what i was doing. but any small change required a full rebuild vs updating a single layer or two.

File: stx-rosetta.Dockerfile
Comment (CharlieC3) [reply stx-rosetta.Dockerfile:6]: I'm saying you can delete just this ENV var and the image will work the same as it did before. There's no benefit to declaring an ARG then setting an ENV var to it if the env var won't be used in the final image because it's redundant. If the repo name changes, you'd update the ARG default value.

File: stx-rosetta.Dockerfile
Comment (CharlieC3) [reply stx-rosetta.Dockerfile:69]: That's fine if you want to keep it that way, I don't find it difficult to read with one directive. Also it seems `ENV` directives don't create additional layers anymore either which is nice.

File: stx-rosetta.Dockerfile
Comment (CharlieC3) [reply stx-rosetta.Dockerfile:103]: > but changes to the repo would require a full rebuild since the cache layers would no longer be used.

If you `COPY` only those ruinit files, changes to the repo wouldn't require a full rebuild. That only happens when you write `COPY . .` for example. Either way will work the same, it's just difficult to read or maintain the scripts when they're embedded.

File: stx-rosetta.Dockerfile
Comment (CharlieC3) [reply stx-rosetta.Dockerfile:70]: > any change in the process/dockerfile would invalidate the cache layers and cause a longer rebuild.

Only true if you `COPY` the dockerfile into the image. For `RUN` directives the docker engine only compares the text of the old `RUN` vs the text of the new `RUN` to determine if it can be reused from cache. These are only minor suggestions, so feel free to do what you like.

**Comment by wileyj** [comment]
ping @zone117x

**Comment by zone117x** [comment]
Thanks @wileyj this looks good to me. Do you think we should merge this to master now, or wait until we merge API develop branch to master?

File: stx-rosetta.Dockerfile
Comment (zone117x) [reply stx-rosetta.Dockerfile:69]: Yes, this env var is no longer used anywhere.

**Comment by wileyj** [comment]
> Thanks @wileyj this looks good to me. Do you think we should merge this to master now, or wait until we merge API develop branch to master?

Up to you! I've shared the file from my branch with the folks who needed it, so there's no rush - pinging just to get it off my plate.

**Comment by zone117x** [comment]
> Up to you! I've shared the file from my branch with the folks who needed it, so there's no rush - pinging just to get it off my plate.

Alright, there could be some changes required once develop branch is ready, and that's expected this week. So I think holding off until then makes sense.

**Comment by wileyj** [comment]
> > Up to you! I've shared the file from my branch with the folks who needed it, so there's no rush - pinging just to get it off my plate.
> 
> Alright, there could be some changes required once develop branch is ready, and that's expected this week. So I think holding off until then makes sense.

👍 changed the base branch to `develop` as well.

**Comment by wileyj** [comment]
ping @zone117x

**Comment by zone117x** [comment]
@wileyj the only concern I have about this PR is the move away from pinned versions (of the stacks-node repo at least). 
See this discussion for more context on versioning issues: https://github.com/blockstack/stacks-blockchain-api/discussions/567

I do think this dockerfile change is a step in the right direction, because the hardcoded versions tend to go stale. 

Should we merge this PR with the possibility that this dockerfile could build images with stacks-api and stacks-node versions that are incompatible with each other (when the release cadences don't match up)?

**Comment by wileyj** [comment]
> #567

Interesting...I don't have any opinion directly on that discussion, but also -  nothing prevents us from setting **specific** versions in the Dockerfile vs trying to dynamically retrieve the version through github api. 

The reason I made the Dockerfile this way was to prevent having to manually update the version numbers each time a 
 new release is cut for either repo (stacks-blockchain or stacks-blockchain-api). If that's not preferred, I can roll back that idea to hardcode the versions. 

Based on the discussion you linked, is there a pair of versions you'd like to start with initially?
i.e.
https://github.com/blockstack/stacks-blockchain-api/releases/tag/v0.58.0
https://github.com/blockstack/stacks-blockchain/releases/tag/2.0.11.0.0


Once I have the specific versions I can update the Dockerfile and we can get this off of our plate. 
The caveat being we'll need to ensure we keep the file up to date with current versions.

**Comment by zone117x** [comment]
> Based on the discussion you linked, is there a pair of versions you'd like to start with initially?
> i.e.
> https://github.com/blockstack/stacks-blockchain-api/releases/tag/v0.58.0
> https://github.com/blockstack/stacks-blockchain/releases/tag/2.0.11.0.0

Yes these would be good versions to set the dockerfile to for now 👍 

> The caveat being we'll need to ensure we keep the file up to date with current versions.

Yeah, this is definitely a pain, hopefully something we can figure out in #567 eventually.

**Review by zone117x — APPROVED** [review]

**Pull request merged**
